File: hospital-frontend/src/features/admin/services/adminService.ts
/**
 * @module features/admin/services/adminService
 *
 * Service layer untuk fitur admin. Semua panggilan lewat API client terpusat
 * (JWT otomatis). Backend membatasi endpoint `/api/admin/**` ke role ADMIN.
 */
import api from "../../../core/api/client";
import type {
  AdminDokter, DokterCreateRequest, DokterUpdateRequest,
  UserAdmin, DashboardStats, Role, PoliRequest, JadwalRequest,
  PasienCreateRequest, AntrianAdmin,
} from "../types";
import type { Poli } from "../../poli/types";
import type { JadwalDokter } from "../../jadwal/types";
import type { Pasien, PasienRequest } from "../../pasien/types";

/* ================= DASHBOARD ================= */

/**
 * Mengambil statistik dashboard admin (GET /api/admin/dashboard).
 *
 * @returns statistik jumlah pasien, dokter, poli, appointment
 */
export const getDashboard = () =>
  api.get<DashboardStats>("/admin/dashboard");

/* ================= MANAJEMEN POLI (FR-35) ================= */

/**
 * Mengambil semua poli termasuk nonaktif (GET /api/admin/poli).
 *
 * @returns daftar semua poli
 */
export const getAllPoliAdmin = () =>
  api.get<Poli[]>("/admin/poli");

/**
 * Membuat poli baru (POST /api/admin/poli).
 *
 * @param payload data poli baru (nama, deskripsi)
 * @returns poli yang dibuat (HTTP 201)
 */
export const createPoli = (payload: PoliRequest) =>
  api.post<Poli>("/admin/poli", payload);

/**
 * Memperbarui poli (PUT /api/admin/poli/{id}).
 *
 * @param id ID poli
 * @param payload data poli baru
 * @returns poli setelah diperbarui
 */
export const updatePoli = (id: number, payload: PoliRequest) =>
  api.put<Poli>(`/admin/poli/${id}`, payload);

/**
 * Menonaktifkan poli (DELETE /api/admin/poli/{id}, soft delete).
 *
 * @param id ID poli
 */
export const deactivatePoli = (id: number) =>
  api.delete(`/admin/poli/${id}`);

/* ================= MANAJEMEN DOKTER (FR-36) ================= */

/**
 * Mengambil semua dokter termasuk nonaktif (GET /api/admin/dokter).
 *
 * @returns daftar semua dokter
 */
export const getAllDokterAdmin = () =>
  api.get<AdminDokter[]>("/admin/dokter");

/**
 * Membuat user dokter + profil dokter sekaligus (POST /api/admin/dokter).
 *
 * @param payload data user dokter baru (nama, email, password, poliId, dll)
 * @returns dokter yang dibuat (HTTP 201)
 */
export const createDokter = (payload: DokterCreateRequest) =>
  api.post<AdminDokter>("/admin/dokter", payload);

/**
 * Memperbarui profil dokter (PUT /api/admin/dokter/{id}).
 *
 * @param id ID dokter
 * @param payload data dokter baru
 * @returns dokter setelah diperbarui
 */
export const updateDokter = (id: number, payload: DokterUpdateRequest) =>
  api.put<AdminDokter>(`/admin/dokter/${id}`, payload);

/**
 * Menonaktifkan dokter (DELETE /api/admin/dokter/{id}, soft delete).
 *
 * @param id ID dokter
 */
export const deactivateDokter = (id: number) =>
  api.delete(`/admin/dokter/${id}`);

/* ================= MANAJEMEN JADWAL (FR-36) ================= */

/**
 * Mengambil semua jadwal (GET /api/admin/jadwal).
 *
 * @returns daftar semua jadwal
 */
export const getAllJadwalAdmin = () =>
  api.get<JadwalDokter[]>("/admin/jadwal");

/**
 * Membuat jadwal baru (POST /api/admin/jadwal).
 *
 * @param payload data jadwal baru
 * @returns jadwal yang dibuat (HTTP 201)
 */
export const createJadwal = (payload: JadwalRequest) =>
  api.post<JadwalDokter>("/admin/jadwal", payload);

/**
 * Memperbarui jadwal (PUT /api/admin/jadwal/{id}).
 *
 * @param id ID jadwal
 * @param payload data jadwal baru
 * @returns jadwal setelah diperbarui
 */
export const updateJadwal = (id: number, payload: JadwalRequest) =>
  api.put<JadwalDokter>(`/admin/jadwal/${id}`, payload);

/**
 * Menonaktifkan jadwal (DELETE /api/admin/jadwal/{id}, soft delete).
 *
 * @param id ID jadwal
 */
export const deactivateJadwal = (id: number) =>
  api.delete(`/admin/jadwal/${id}`);

/* ================= MANAJEMEN PENGGUNA (FR-04) ================= */

/**
 * Mengambil daftar semua pengguna, opsional difilter per role
 * (GET /api/admin/users?role=...).
 *
 * @param role filter role (opsional)
 * @returns daftar pengguna dengan status aktif
 */
export const getAllUsers = (role?: Role) =>
  api.get<UserAdmin[]>("/admin/users", { params: role ? { role } : {} });

/**
 * Mengaktifkan/nonaktifkan akun pengguna (PATCH /api/admin/users/{id}/status).
 *
 * @param id ID user
 * @param active status aktif baru
 * @returns user setelah diperbarui
 */
export const setUserStatus = (id: number, active: boolean) =>
  api.patch<UserAdmin>(`/admin/users/${id}/status`, null, { params: { active } });

/**
 * Reset password pengguna (PATCH /api/admin/users/{id}/password).
 *
 * Admin mengganti password dokter/pasien tanpa perlu tahu password lama.
 *
 * @param id ID user
 * @param password password baru (min 6 karakter)
 * @returns user setelah diperbarui
 */
export const resetPassword = (id: number, password: string) =>
  api.patch<UserAdmin>(`/admin/users/${id}/password`, { password });

/* ================= MANAJEMEN PASIEN ================= */

/**
 * Mengambil daftar semua pasien (GET /api/admin/pasien).
 *
 * @returns daftar semua pasien
 */
export const getAllPasienAdmin = () =>
  api.get<Pasien[]>("/admin/pasien");

/**
 * Membuat user pasien + profil pasien sekaligus (POST /api/admin/pasien).
 *
 * @param payload data user pasien baru (nama, email, password, nik, dll)
 * @returns pasien yang dibuat (HTTP 201)
 */
export const createPasien = (payload: PasienCreateRequest) =>
  api.post<Pasien>("/admin/pasien", payload);

/**
 * Memperbarui profil pasien (PUT /api/admin/pasien/{id}).
 *
 * @param id ID pasien
 * @param payload data profil pasien
 * @returns pasien setelah diperbarui
 */
export const updatePasien = (id: number, payload: PasienRequest) =>
  api.put<Pasien>(`/admin/pasien/${id}`, payload);

/**
 * Menonaktifkan akun pasien (DELETE /api/admin/pasien/{id}).
 *
 * @param id ID pasien
 */
export const deactivatePasien = (id: number) =>
  api.delete(`/admin/pasien/${id}`);

/* ================= MANAJEMEN ANTRIAN ================= */

/**
 * Mengambil semua antrian untuk papan antrian admin (GET /api/admin/antrian).
 *
 * @returns daftar semua antrian
 */
export const getAllAntrianAdmin = () =>
  api.get<AntrianAdmin[]>("/admin/antrian");

File: medikasentosa/src/main/java/com/medikasentosa/features/admin/service/AdminService.java
package com.medikasentosa.features.admin.service;

import com.medikasentosa.features.admin.dto.DashboardResponse;
import com.medikasentosa.features.admin.dto.DokterCreateRequest;
import com.medikasentosa.features.admin.dto.PasienCreateRequest;
import com.medikasentosa.features.admin.dto.UserAdminResponse;
import com.medikasentosa.features.appointment.repository.AppointmentRepository;
import com.medikasentosa.features.auth.entity.Role;
import com.medikasentosa.features.auth.entity.User;
import com.medikasentosa.features.auth.repository.UserRepository;
import com.medikasentosa.features.dokter.dto.DokterRequest;
import com.medikasentosa.features.dokter.dto.DokterResponse;
import com.medikasentosa.features.dokter.entity.Dokter;
import com.medikasentosa.features.dokter.repository.DokterRepository;
import com.medikasentosa.features.dokter.service.DokterService;
import com.medikasentosa.features.pasien.dto.PasienRequest;
import com.medikasentosa.features.pasien.dto.PasienResponse;
import com.medikasentosa.features.pasien.service.PasienService;
import com.medikasentosa.features.poli.repository.PoliRepository;
import com.medikasentosa.shared.exception.DuplicateResourceException;
import com.medikasentosa.shared.exception.ResourceNotFoundException;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

/**
 * Service untuk operasi administratif (FR-35, FR-36, FR-04).
 *
 * Mengelola: pembuatan user dokter + profil dokter dalam satu transaksi,
 * daftar semua pengguna, aktivasi/nonaktifkan akun, dan statistik dashboard.
 * Akses dibatasi role ADMIN lewat {@code @PreAuthorize} pada controller.
 *
 * @author Ari
 * @since 1.0.0
 */
@Service
public class AdminService {

    private final UserRepository userRepository;
    private final DokterRepository dokterRepository;
    private final PoliRepository poliRepository;
    private final AppointmentRepository appointmentRepository;
    private final DokterService dokterService;
    private final PasienService pasienService;
    private final PasswordEncoder passwordEncoder;

    public AdminService(UserRepository userRepository,
                        DokterRepository dokterRepository,
                        PoliRepository poliRepository,
                        AppointmentRepository appointmentRepository,
                        DokterService dokterService,
                        PasienService pasienService,
                        PasswordEncoder passwordEncoder) {
        this.userRepository = userRepository;
        this.dokterRepository = dokterRepository;
        this.poliRepository = poliRepository;
        this.appointmentRepository = appointmentRepository;
        this.dokterService = dokterService;
        this.pasienService = pasienService;
        this.passwordEncoder = passwordEncoder;
    }

    /**
     * Membuat user dokter baru beserta profil dokternya dalam satu transaksi (FR-36).
     *
     * Menggabungkan pembuatan akun (User role=DOKTER) dan profil dokter sehingga
     * admin tidak perlu mendaftar user terpisah. Bila email sudah dipakai, lempar
     * 409. Bila user sudah terdaftar sebagai dokter, lempar 409.
     *
     * @param request data user dokter baru (nama, email, password, poliId, noStr, dll)
     * @return data dokter yang berhasil dibuat
     * @throws DuplicateResourceException bila email sudah terdaftar
     * @throws ResourceNotFoundException  bila poli tidak ditemukan
     */
    @Transactional
    public DokterResponse createDokterWithUser(DokterCreateRequest request) {
        // Validasi email unik
        if (userRepository.existsByEmail(request.email())) {
            throw new DuplicateResourceException("Email sudah terdaftar");
        }

        // Buat user baru dengan role DOKTER
        User user = User.builder()
                .nama(request.nama())
                .email(request.email())
                .password(passwordEncoder.encode(request.password()))
                .role(Role.DOKTER)
                .isActive(true)
                .build();
        user = userRepository.save(user);

        // Buat profil dokter via DokterService (sudah handle validasi poli & duplikat)
        DokterRequest dokterReq = new DokterRequest(
                user.getId(), request.poliId(), request.noStr(),
                request.spesialisasi(), request.noHp());
        return dokterService.create(dokterReq);
    }

    /**
     * Mengambil daftar dokter termasuk yang nonaktif (untuk admin).
     *
     * Berbeda dengan {@code DokterService.getAll()} yang hanya mengembalikan
     * dokter aktif, method ini menampilkan semua dokter agar admin bisa
     * melihat yang sudah dinonaktifkan.
     *
     * @return daftar semua dokter (aktif & nonaktif)
     */
    public List<DokterResponse> getAllDokterForAdmin() {
        return dokterRepository.findAll().stream()
                .map(this::toDokterResponse)
                .toList();
    }

    /**
     * Membuat user pasien baru beserta profil pasiennya dalam satu transaksi.
     *
     * Menggabungkan pembuatan akun (User role=PASIEN) dan profil pasien (NIK,
     * alamat, dll) sehingga admin bisa menambah pasien manual tanpa minta pasien
     * daftar sendiri via /register.
     *
     * @param request data user pasien baru (nama, email, password, nik, dll)
     * @return data pasien yang berhasil dibuat
     * @throws DuplicateResourceException bila email sudah terdaftar
     */
    @Transactional
    public PasienResponse createPasienWithUser(PasienCreateRequest request) {
        if (userRepository.existsByEmail(request.email())) {
            throw new DuplicateResourceException("Email sudah terdaftar");
        }

        User user = User.builder()
                .nama(request.nama())
                .email(request.email())
                .password(passwordEncoder.encode(request.password()))
                .role(Role.PASIEN)
                .isActive(true)
                .build();
        user = userRepository.save(user);

        PasienRequest pasienReq = new PasienRequest(
                user.getId(), request.nik(), request.tanggalLahir(),
                request.jenisKelamin(), request.alamat(), request.noHp(),
                request.golDarah());
        return pasienService.create(pasienReq);
    }

    /**
     * Memperbarui profil pasien (dipakai admin untuk edit data pasien).
     *
     * @param id      ID pasien
     * @param request data profil pasien baru
     * @return data pasien setelah diperbarui
     */
    public PasienResponse updatePasien(Long id, PasienRequest request) {
        return pasienService.update(id, request);
    }

    /**
     * Mengambil daftar semua pengguna, opsional difilter per role (FR-04 implisit).
     *
     * @param role role untuk filter (boleh null = semua role)
     * @return daftar pengguna dengan status aktif
     */
    public List<UserAdminResponse> getAllUsers(Role role) {
        List<User> users = (role == null) ? userRepository.findAll() : userRepository.findByRole(role);
        return users.stream().map(this::toUserAdminResponse).toList();
    }

    /**
     * Mengaktifkan atau menonaktifkan akun pengguna (FR-04 implisit).
     *
     * Akun nonaktif tidak dapat login (ditolak di AuthService.login).
     *
     * @param id      ID user yang akan diubah statusnya
     * @param active  status aktif baru (true = aktifkan, false = nonaktifkan)
     * @throws ResourceNotFoundException bila user tidak ditemukan
     */
    @Transactional
    public UserAdminResponse setUserStatus(Long id, boolean active) {
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User tidak ditemukan"));
        user.setIsActive(active);
        userRepository.save(user);
        return toUserAdminResponse(user);
    }

    /**
     * Mengganti password pengguna (admin reset password dokter/pasien).
     *
     * Password baru di-hash dengan BCrypt sebelum disimpan. Admin tidak
     * perlu tahu password lama — ini fitur reset, bukan change password mandiri.
     *
     * @param id           ID user yang akan diganti passwordnya
     * @param newPassword  password baru (plain text, akan di-hash)
     * @throws ResourceNotFoundException bila user tidak ditemukan
     * @throws IllegalArgumentException  bila password baru kosong/terlalu pendek
     */
    @Transactional
    public UserAdminResponse resetPassword(Long id, String newPassword) {
        if (newPassword == null || newPassword.length() < 6) {
            throw new IllegalArgumentException("Password baru minimal 6 karakter");
        }
        User user = userRepository.findById(id)
                .orElseThrow(() -> new ResourceNotFoundException("User tidak ditemukan"));
        user.setPassword(passwordEncoder.encode(newPassword));
        userRepository.save(user);
        return toUserAdminResponse(user);
    }

    /**
     * Mengambil statistik ringkas untuk dashboard admin.
     *
     * @return jumlah pasien, dokter, poli, dan appointment
     */
    public DashboardResponse getDashboard() {
        long totalPasien = userRepository.countByRole(Role.PASIEN);
        long totalDokter = dokterRepository.count();
        long totalPoli = poliRepository.count();
        long totalAppointment = appointmentRepository.count();
        return new DashboardResponse(totalPasien, totalDokter, totalPoli, totalAppointment);
    }

    private UserAdminResponse toUserAdminResponse(User user) {
        return new UserAdminResponse(
                user.getId(), user.getNama(), user.getEmail(),
                user.getRole(), user.getIsActive());
    }

    private DokterResponse toDokterResponse(Dokter dokter) {
        return new DokterResponse(
                dokter.getId(),
                dokter.getUser().getId(),
                dokter.getUser().getNama(),
                dokter.getUser().getEmail(),
                dokter.getPoli().getId(),
                dokter.getPoli().getNama(),
                dokter.getNoStr(),
                dokter.getSpesialisasi(),
                dokter.getNoHp());
    }
}
